packages/react-compose/src/createClassResolver.ts

```diff
@@ -65,3 +66,59 @@ function addClassTo(slotProps: GenericDictionary, slotName: string, className?:
     }
   }
 }
+
+type ResolvedMap = {
+  slots: Record<string, string | string[]>;
+  modifiers: Record<string, string | string[]>;
+  enums: Record<string, Record<string, string | string[]>>;
+};
+
+/**
+ * Helper to take a css module map and translate it into { slots, modifiers,  enums } where
+ * slots are a matched name in the slotNames array, enums have underscores splitting the matched
+ * name/value, and modifiers are everything else. Creating this split definition keeps runtime
+ * resolution work to a minimum.
+ */
+function createResolvedMap(classes: ClassDictionary, slotNames: string[]): ResolvedMap {
+  const resolvedMap: ResolvedMap = {
+    slots: {},
+    modifiers: {},
+    enums: {},
+  };
+  const { slots, modifiers, enums } = resolvedMap;
+
+  // Iterate through classes
+  Object.keys(classes).forEach((key: string) => {
+    const classValue = classes[key];
+
+    if (classValue) {
+      // If the class is named the same as a slot, add it to the slot.
+      if (slotNames.indexOf(key) >= 0) {
+        addClassTo(slots, key, classValue);
+      } else if (key.indexOf('_') >= 0) {
+        // The class is an enum value. Add if the prop exists and matches.
+        const parts = key.split('_');
+        const enumName = parts[0];
+        const enumValue = parts[1];
+
+        enums[enumName] = enums[enumName] || {};
+        addClassTo(enums[enumName], enumValue, classValue);
+      } else {
+        addClassTo(modifiers, key, classValue);
+      }
+    }
+  });
+
+  // Reduce map to strings
+  collapseArray(slots);
+  collapseArray(modifiers);
```

Comment:
instead of building array then join. would it be better if we just start with string and append the classname in `addClassTo`